### PR TITLE
sigstore, test: honor PublicKeyDetails when loading Keyrings

### PR DESCRIPTION
**Pull request by woodruffw**
See https://github.com/sigstore/sigstore-conformance/issues/140 for some context -- we were previously ignoring `PublicKeyDetails` entirely.

This adds a `Key` abstraction that gets constructed from the `PublicKey` message, handling `PublicKeyDetails` in the process. It also cleans up some of our error types and old tests (like `test_ctfe.py`, which is entirely obviated by the trustroot APIs and their tests now).

Separately, this will also "fix" our handling of key IDs per https://github.com/sigstore/rekor/issues/2062, in the sense that we'll go from failing closed on key ID mismatch to trying all keys in the keyring.

### CODE REVIEW

File: sigstore/_internal/trustroot.py [hunk 1]
```diff
@@ -102,72 +106,116 @@ class KeyringSignatureError(KeyringError):
...
+        if public_key.key_details in self._RSA_SHA_256_DETAILS:
+            hash_algorithm = hashes.SHA256()
+            key = load_der_public_key(public_key.raw_bytes, types=(rsa.RSAPublicKey,))
+        elif public_key.key_details in self._EC_DETAILS_TO_HASH:
+            hash_algorithm = self._EC_DETAILS_TO_HASH[public_key.key_details]
+            key = load_der_public_key(
+                public_key.raw_bytes, types=(ec.EllipticCurvePublicKey,)
+            )
+        else:
+            raise InvalidKeyError(f"unsupported key type: {public_key.key_details}")
```
Comment (woodruffw) [inline sigstore/_internal/trustroot.py:147]: NB: This brings us to parity with the previous behavior, i.e. allowing ECDSA and RSA. But it gives us the flexility to allow additional keytypes in the future 🙂

File: test/unit/internal/test_ctfe.py
Comment (woodruffw) [inline test/unit/internal/test_ctfe.py:1]: NB: This file entirely tested the old keyring loading APIs, including APIs that were never used anyways (like `Keyring.add`)

File: sigstore/_internal/trustroot.py (hunk 1 above)
Comment (haydentherapper) [reply sigstore/_internal/trustroot.py:147]: +1 to a nice error message

File: sigstore/_internal/trustroot.py [hunk 2]
```diff
@@ -102,72 +106,119 @@ class KeyringSignatureError(KeyringError):
...
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_3072_SHA256,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_4096_SHA256,
+    }
+
+    _EC_DETAILS_TO_HASH: ClassVar[dict[_PublicKeyDetails, hashes.HashAlgorithm]] = {
+        _PublicKeyDetails.PKIX_ECDSA_P256_SHA_256: hashes.SHA256(),
+        _PublicKeyDetails.PKIX_ECDSA_P384_SHA_384: hashes.SHA384(),
+        _PublicKeyDetails.PKIX_ECDSA_P521_SHA_512: hashes.SHA512(),
+    }
+
```
Comment (haydentherapper) [inline sigstore/_internal/trustroot.py:131]: no need to do this here, but is it relatively straightforward to add ed25519 down the line?

**Review by haydentherapper — COMMENTED** [review]
> Separately, this will also "fix" our handling of key IDs per https://github.com/sigstore/rekor/issues/2062, in the sense that we'll go from failing closed on key ID mismatch to trying all keys in the keyring.

Only concern with this is it's inefficient to verify with all keys. Is the motivation to avoid dealing with the key ID calculation and continuing to support RSA?

File: sigstore/_internal/trustroot.py (hunk 2 above)
Comment (woodruffw) [reply sigstore/_internal/trustroot.py:131]: Yep! It should just be another branch (below) in the future.

**Comment by woodruffw** [comment]
> Only concern with this is it's inefficient to verify with all keys. Is the motivation to avoid dealing with the key ID calculation and continuing to support RSA?

Pretty much -- I intend to fix the key ID calculation as well, but I didn't want to blow this diff up too much 🙂

**Review by jku — APPROVED** [review]
looks good, thanks. 

test_trust_root.py is really begging to be simplified a bit... but this is fine.

**Pull request merged**